src/piton/labelers/core.py

```diff
@@ -0,0 +1,381 @@
+    """
+
+    __slots__ = [
+        "time", # Arbitrary timestamp (datetime.datetime)
+        "label_type",
+        "value",
+    ]
+
+    label_type: LabelType
+
+    def __init__(
+        self,
+        time: datetime.datetime,
+        value: Union[bool, int, float, SurvivalValue],
+        label_type: LabelType,
+    ):
+        """Construct a label for datetime `time` and value `value`. 
+
+        Args:
+            time (datetime.datetime): Time in this patient's timeline that corresponds to this label
+            value (Union[bool, int, float, SurvivalValue]): Value of label. Defaults to None.
+            label_type (LabelType): Type of label. Must be an element in `VALID_LABEL_TYPES`.
+        """
+        assert label_type in VALID_LABEL_TYPES
+        self.time = time
+        self.label_type = label_type
+        self.value = value
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Label):
+            return NotImplemented
+        return (
+            self.time,
+            self.label_type,
+            self.value,
+        ) == (
+            other.time,
+            other.label_type,
+            other.value,
+        )
+
+    def __repr__(self) -> str:
+        return f"Label(time={self.time}, value={self.value}, type={self.label_type})"
+
+class LabelingFunction(ABC):
+    """
+    An interface for labeling functions.
+    
+    A labeling function applies a label to a specific datetime in a given patient's timeline.
+    It can be thought of as generating the following list given a specific patient:
+        [(patient ID, datetime_1, label_1), (patient ID, datetime_2, label_2), ... ]
+    
+    Usage:
+        labeling_function: LabelingFunction = LF(...)
+        patients: Sequence[Patient] = ...
+        labels: LabeledPatient = labeling_function.apply(patients)
+    """
+
+    @abstractmethod
+    def label(self, patient: Patient) -> List[Label]:
+        """
+        Applies every label that is applicable to the provided patient.
+        This is only called once per patient.
+
+        Args:
+            patient (Patient): A patient object
+
+        Returns:
+            List[Label]: A list of :class:`Label` containing every label for the given patient
+        """
+        pass
+
+    def get_required_codes(self) -> List[int]:
+        """Set of codes that a patient must have at least one of in order to qualify for this labeler
+        This allows us to only extract patients from the :class:`PatientDatabase` who have a code that matches one of these
+
+        Returns:
+            List[int]: List of applicable OMOP codes
+        """
+        pass
+
+    def get_patient_start_end_times(self, patient: Patient) -> Tuple[datetime.datetime, datetime.datetime]:
+        """Returns the (start, end) of the patient timeline
+
+        Returns:
+            Tuple[datetime.datetime, datetime.datetime]: (start, end)
+        """
+        pass
```

Comment:
As noted before, I think this can be removed, but we won't know for sure until later.

In the meantime, please add a TODO: Evaluate whether this can be removed as part of the docstring of this function.